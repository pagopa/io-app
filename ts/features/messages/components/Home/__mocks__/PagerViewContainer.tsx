import React from "react";
import { View } from "react-native";
import PagerView from "react-native-pager-view";

export const PagerViewContainer = React.forwardRef<PagerView>((_, ref) => (
  <PagerView ref={ref} testID="mock_pager_view_container">
    <View>Mock Inbox</View>
    <View>Mock Archive</View>
  </PagerView>
));
