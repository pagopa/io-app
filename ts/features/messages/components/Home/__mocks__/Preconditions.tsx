import React from "react";
import { View } from "react-native";

export const Preconditions = () => (
  <View testID="mock_preconditions">This is a mock for Preconditions</View>
);
