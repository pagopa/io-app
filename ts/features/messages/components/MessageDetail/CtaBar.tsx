import * as O from "fp-ts/lib/Option";
import React, { useEffect, useRef } from "react";
import { View, StyleSheet } from "react-native";
import { useStore } from "react-redux";
import { HSpacer } from "@pagopa/io-app-design-system";
import { CommonServiceMetadata } from "../../../../../definitions/backend/CommonServiceMetadata";
import { ServiceId } from "../../../../../definitions/backend/ServiceId";
import { useIODispatch } from "../../../../store/hooks";
import { PaymentData, UIMessageDetails, UIMessageId } from "../../types";
import { UIService } from "../../../../store/reducers/entities/services/types";
import variables from "../../../../theme/variables";

import {
  getMessageCTA,
  isExpired,
  MessagePaymentExpirationInfo
} from "../../utils/messages";
import { IOStyles } from "../../../../components/core/variables/IOStyles";
import ExtractedCTABar from "../../../../components/cta/ExtractedCTABar";
import { isPNOptInMessage } from "../../../pn/utils";
import {
  trackPNOptInMessageCTADisplaySuccess,
  trackPNOptInMessageOpened
} from "../../../pn/analytics";
import { getPaymentExpirationInfo } from "../../utils";
import CalendarEventButton from "./CalendarEventButton";
import PaymentButton from "./PaymentButton";

type Props = {
  isPaid: boolean;
  messageDetails: UIMessageDetails;
  service?: UIService;
  serviceMetadata?: CommonServiceMetadata;
};

const styles = StyleSheet.create({
  footerContainer: {
    overflow: "hidden",
    marginTop: -variables.footerShadowOffsetHeight,
    paddingTop: variables.footerShadowOffsetHeight
  }
});

// return a payment button only when the advice is not paid and the payment_data is defined
const renderPaymentButton = (
  paymentData?: PaymentData,
  messageId?: UIMessageId
) => {
  if (paymentData === undefined) {
    return null;
  }
  const { amount, noticeNumber, payee } = paymentData;
  return (
    <PaymentButton
      amount={amount}
      noticeNumber={noticeNumber}
      organizationFiscalCode={payee.fiscalCode}
      messageId={messageId}
    />
  );
};

function renderCalendarEventButton(
  isPaid: boolean,
  expirationInfo: MessagePaymentExpirationInfo,
  legacyMessage: UIMessageDetails["raw"],
  dueDate?: Date
) {
  if (dueDate === undefined || isPaid || isExpired(expirationInfo)) {
    return null;
  }
  return <CalendarEventButton message={legacyMessage} />;
}

/**
 * A smart component to show the action buttons of a message.
 * It requires context and full redux support.
 *
 * For messages with the proper configuration renders, on a row:
 * - a calendar icon
 * - a button to add/remove a calendar event
 * - a button to show/start a payment
 */
const CtaBar = ({
  isPaid,
  messageDetails,
  service,
  serviceMetadata
}: Props): React.ReactElement | null => {
  const dispatch = useIODispatch();
  const shoulCheckForPNOptInMessage = useRef(true);
  const store = useStore();

  const { dueDate, markdown, paymentData, raw: legacyMessage } = messageDetails;
  const maybeCtas = getMessageCTA(
    markdown,
    serviceMetadata,
    service?.id as ServiceId
  );
  const state = store.getState();
  const isPNOptInMessageInfo = isPNOptInMessage(maybeCtas, service, state);
  const isPNOptIn = isPNOptInMessageInfo.isPNOptInMessage;

  useEffect(() => {
    if (shoulCheckForPNOptInMessage.current && isPNOptIn) {
      // eslint-disable-next-line functional/immutable-data
      shoulCheckForPNOptInMessage.current = false;
      trackPNOptInMessageOpened();
      trackPNOptInMessageCTADisplaySuccess();
    }
  }, [isPNOptIn, shoulCheckForPNOptInMessage]);

  const expirationInfo = getPaymentExpirationInfo(messageDetails);

  const paymentButton = renderPaymentButton(paymentData, messageDetails.id);
  const calendarButton = renderCalendarEventButton(
    isPaid,
    expirationInfo,
    legacyMessage,
    dueDate
  );

  const footer1 = (paymentButton || calendarButton) && (
    // Added a wrapper to enable the usage of the component outside the Container of Native Base
    <View style={styles.footerContainer} pointerEvents={"box-none"}>
      <View style={[IOStyles.footer, IOStyles.row]}>
        {calendarButton}
        {paymentButton && calendarButton && <HSpacer size={16} />}
        {paymentButton}
      </View>
    </View>
  );

  const footer2 = O.isSome(maybeCtas) && (
    // Added a wrapper to enable the usage of the component outside the Container of Native Base
    <View style={styles.footerContainer} pointerEvents={"box-none"}>
      <View testID={"CtaBar_withCTA"} style={[IOStyles.footer, IOStyles.row]}>
        <ExtractedCTABar
          ctas={maybeCtas.value}
          xsmall={false}
          dispatch={dispatch}
          serviceMetadata={serviceMetadata}
          service={service?.raw}
          isPNOptInMessage={isPNOptInMessageInfo}
        />
      </View>
    </View>
  );
  return (
    <View testID={"CtaBar_withFooter"}>
      {footer2}
      {footer1}
    </View>
  );
};

export default CtaBar;
