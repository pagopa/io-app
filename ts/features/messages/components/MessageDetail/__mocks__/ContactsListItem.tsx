import { View } from "react-native";
import { ContactsListItemProps } from "../ContactsListItem";

export const ContactsListItem = ({ email, phone }: ContactsListItemProps) => (
  <>
    <View>Mock ContactsListItem</View>
    {email && <View>{email}</View>}
    {phone && <View>{phone}</View>}
  </>
);
