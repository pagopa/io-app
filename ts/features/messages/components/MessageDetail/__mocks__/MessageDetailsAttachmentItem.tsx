import { Pressable, View } from "react-native";
import { MessageDetailsAttachmentItemProps } from "../MessageDetailsAttachmentItem";

export const MessageDetailsAttachmentItem = ({
  attachment,
  bottomSpacer,
  disabled,
  isPN,
  messageId,
  onPreNavigate,
  serviceId
}: MessageDetailsAttachmentItemProps) => (
  <>
    <View>Mock MessageDetailsAttachmentItem</View>
    <View>
      <View>{`Attachment category:     ${attachment.category}`}</View>
      <View>{`Attachment content-type: ${attachment.content_type}`}</View>
      <View>{`Attachment id:           ${attachment.id}`}</View>
      <View>{`Attachment name:         ${attachment.name}`}</View>
      <View>{`Attachment url:          ${attachment.url}`}</View>
    </View>
    <View>{`Has${bottomSpacer ? "" : " no"} bottom spacer`}</View>
    <View>{`Is ${disabled ? "" : "not "}disabled`}</View>
    <View>{`Is ${isPN ? "" : "not "}SEND`}</View>
    <View>{`Message Id: ${messageId}`}</View>
    <View>{`Service Id: ${serviceId}`}</View>
    <Pressable accessibilityLabel="" onPress={onPreNavigate}>
      {`Has${onPreNavigate != null ? "" : " no"} onPreNavigate callback`}
    </Pressable>
  </>
);
