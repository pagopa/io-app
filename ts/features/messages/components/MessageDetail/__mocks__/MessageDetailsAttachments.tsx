import { View } from "react-native";
import { MessageDetailsAttachmentsProps } from "../MessageDetailsAttachments";

export const MessageDetailsAttachments = ({
  banner,
  disabled,
  isPN,
  messageId,
  serviceId
}: MessageDetailsAttachmentsProps) => (
  <>
    <View>Mock MessageDetailsAttachments</View>
    <View>{`${banner ? "Has" : "Does not have"} Banner`}</View>
    <View>{`Is ${disabled ? "" : "not "}disabled`}</View>
    <View>{`Is ${isPN ? "" : "not "}SEND`}</View>
    <View>{`Message Id: ${messageId}`}</View>
    <View>{`Service Id: ${serviceId}`}</View>
  </>
);
