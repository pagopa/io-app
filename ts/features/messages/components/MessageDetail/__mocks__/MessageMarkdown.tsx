import React from "react";
import WebView from "react-native-webview";
import { MarkdownProps } from "../../../../../components/ui/Markdown/Markdown";

type Props = Omit<MarkdownProps, "cssStyle">;

export const MessageMarkdown = ({ onError, onLoadEnd, testID }: Props) => (
  <WebView onError={onError} onLoadEnd={onLoadEnd} testID={testID}>
    Mock Message Markdown
  </WebView>
);
