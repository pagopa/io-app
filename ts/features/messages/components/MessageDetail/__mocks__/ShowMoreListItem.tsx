import { View } from "react-native";
import { ShowMoreListItemProps } from "../ShowMoreListItem";

export const ShowMoreListItem = ({ sections }: ShowMoreListItemProps) => (
  <>
    <View>Mock ShowMoreListItem</View>
    {sections.map((section, index) => (
      <View key={index}>
        <View>{`View at ${index}`}</View>
        <View>{section.title}</View>
        {section.items.map((item, itemIndex) => (
          <View key={itemIndex}>
            <View>{`item at ${itemIndex}`}</View>
            <View>{item.accessibilityLabel}</View>
            {item.icon && <View>{item.icon}</View>}
            <View>{item.label}</View>
            <View>{item.value}</View>
            {item.valueToCopy && <View>{item.valueToCopy}</View>}
          </View>
        ))}
      </View>
    ))}
  </>
);
