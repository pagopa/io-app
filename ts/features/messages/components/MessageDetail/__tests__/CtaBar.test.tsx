import React from "react";

import configureMockStore from "redux-mock-store";
import I18n from "../../../../../i18n";
import { applicationChangeState } from "../../../../../store/actions/application";
import { appReducer } from "../../../../../store/reducers";
import { paymentValidInvalidAfterDueDate } from "../../../__mocks__/message";
import { service_1 } from "../../../__mocks__/messages";

import { toUIMessageDetails } from "../../../store/reducers/transformers";
import { toUIService } from "../../../../../store/reducers/entities/services/transformers";
import { GlobalState } from "../../../../../store/reducers/types";
import { renderScreenWithNavigationStoreContext } from "../../../../../utils/testWrapper";

import CtaBar from "../CtaBar";
import { MESSAGES_ROUTES } from "../../../navigation/routes";

jest.useFakeTimers();

const uiMessageDetails = toUIMessageDetails(paymentValidInvalidAfterDueDate);
const uiService = toUIService(service_1);

const defaultProps = {
  isPaid: true,
  messageDetails: uiMessageDetails,
  service: uiService,
  servicesMetadata: undefined
};

describe("the `CtaBar` component", () => {
  describe("when neither `paymentData` nor `dueDate` are defined for the message details", () => {
    it("should match the snapshot (no buttons)", () => {
      const noPaymentMessage = {
        ...uiMessageDetails,
        paymentData: undefined,
        dueDate: undefined
      };
      const { component } = renderComponent({
        ...defaultProps,
        messageDetails: noPaymentMessage
      });
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe("when `paymentData` is defined", () => {
    describe("and `isPaid` is true", () => {
      it("should render the payment button", () => {
        const { component } = renderComponent(defaultProps);
        expect(
          component.queryByText(I18n.t("messages.cta.seeNotice"))
        ).not.toBeNull();
      });
    });
    describe("and `isPaid` is false", () => {
      it("should render the payment button", () => {
        const { component } = renderComponent({
          ...defaultProps,
          isPaid: false
        });
        expect(
          component.queryByText(I18n.t("messages.cta.seeNotice"))
        ).not.toBeNull();
      });
    });
  });

  describe("when `dueDate` is defined", () => {
    describe("and `isPaid` is true", () => {
      it("should not render the calendar button", () => {
        const { component } = renderComponent(defaultProps);
        expect(
          component.queryByText(I18n.t("messages.cta.reminder"))
        ).toBeNull();
      });
    });
    describe("and `isPaid` is false", () => {
      it("should render the calendar button", () => {
        const { component } = renderComponent({
          ...defaultProps,
          messageDetails: {
            ...defaultProps.messageDetails,
            dueDate: new Date(new Date().getTime() + 1000 * 60 * 60 * 24)
          },
          isPaid: false
        });
        expect(
          component.queryByText(I18n.t("messages.cta.reminder"))
        ).not.toBeNull();
      });
    });
    describe("and the payment has expired", () => {
      it("should not render the calendar button", () => {
        const { component } = renderComponent({
          ...defaultProps
        });
        expect(
          component.queryByText(I18n.t("messages.cta.reminder"))
        ).toBeNull();
      });
    });
  });

  describe("when at least one CTA is defined in the markdown", () => {
    const markdown = `---
en:
    cta_1: 
        text: "DON'T PANIC!"
        action: "ioit://PROFILE_MAIN"
---
                      ` as any;

    it("should render it", () => {
      const { component } = renderComponent({
        ...defaultProps,
        messageDetails: {
          ...defaultProps.messageDetails,
          markdown
        }
      });
      expect(component.queryByText("DON'T PANIC!")).not.toBeNull();
    });
  });
});

const renderComponent = (props: React.ComponentProps<typeof CtaBar>) => {
  const globalState = appReducer(undefined, applicationChangeState("active"));
  const mockStore = configureMockStore<GlobalState>();
  const store: ReturnType<typeof mockStore> = mockStore(globalState);

  return {
    component: renderScreenWithNavigationStoreContext<GlobalState>(
      () => <CtaBar {...props} />,
      MESSAGES_ROUTES.MESSAGE_DETAIL,
      {},
      store
    ),
    store
  };
};
