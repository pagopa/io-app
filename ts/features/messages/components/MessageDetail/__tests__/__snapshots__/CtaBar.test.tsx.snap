// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the \`CtaBar\` component when neither \`paymentData\` nor \`dueDate\` are defined for the message details should match the snapshot (no buttons) 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  MESSAGE_DETAIL
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSModalScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            Object {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          onGestureCancel={[Function]}
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        testID="CtaBar_withFooter"
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSModalScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
